README.txt
*************************************************
* LibSerializer: Version 0.3.0                  *
*************************************************

30 May 2008

(C)opyright, 2006-2008, by Pentaho Corporation, Object Refinery Limited and Contributors.

-----------------
1.  INTRODUCTION
-----------------
This library module contains a general serialization framework. It simplifies
the task of writing custom serialization handlers for non-serializable classes.

LibSerializer is licensed, free of charge, under the terms of the GNU Lesser
General Public Licence.  A copy of the licence is included in the download.

For the latest news and information about LibSerializer, please refer to:

    http://reporting.pentaho.org/libserializer/


2. REQUIREMENTS
---------------
LibSerializer needs at least JDK 1.2.2 for its functionality.


3. SUPPORT
----------
Free support is available via the Pentaho forum, follow the link
from the home page.  Please note that questions are answered by volunteers, 
so there is no guaranteed response time or level of service.

Please avoid e-mailing the developers directly for support questions.
If you post a message in the forum, then everyone can see the
question, and everyone can see the answer.

The forums can be found at

  http://forums.pentaho.org/


4. REPORTING BUGS
-----------------
If you find bugs in LibSerializer, we'd like to hear about it so that we
can improve future releases of LibSerializer.  Please post a bug report
to the JIRA bug-tracker at Pentaho.org:

    http://jira.pentaho.org/

Please be sure to provide as much information as you can.  We need to
know the version of LibSerializer that you are using, the JDK version,
and the steps required to replicate the bug.  Include any other
information that you think is relevant.


5. ANT
------
We use an open source build tool called Ant to build LibSerializer.  An
Ant script (tested using Ant 1.6) is included in the distribution:

    <libserializer-directory>/build.xml

You can find out more about Ant at:

    http://ant.apache.org/

Ant is licensed under the terms of the Apache Software License (a
popular open source software license).


6. OTHER FEEDBACK
-----------------
For other feedback and comments, please post a message on the
Pentaho forums. The Forum is available at

  http://forums.pentaho.org/



ChangeLog.txt
---------------
1.  WHAT's NEW
---------------
A list of changes in recent versions:

0.3.0: (30-May-2008)
        * Switched from JCommon to LibBase. All version information is now contained in
          the manifest. The Project-Info implementation reads the version numbers from
          the Jar's Manifest or the compiler-output-directory.

0.2.0: (01-Aug-2007)
        * [BUG] The initialization configuration file must be named after the
          library. The 'serializer.properties' file has been renamed into
          'jcommon-serializer.properties'.

0.1.0: (17-Apr-2006)
        * The first public release of this library.



source/org/pentaho/reporting/libraries/serializer/SerializerHelper.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Object Refinery Limited, Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.serializer;

import java.io.IOException;
import java.io.NotSerializableException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.HashMap;
import java.util.Iterator;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.pentaho.reporting.libraries.base.config.Configuration;
import org.pentaho.reporting.libraries.base.util.ObjectUtilities;
import org.pentaho.reporting.libraries.base.util.DebugLog;


/**
 * The SerializeHelper is used to make implementing custom serialization
 * handlers easier. Handlers for certain object types need to be added to this
 * helper before this implementation is usable.
 *
 * @author Thomas Morgner
 */
public class SerializerHelper
{
  private static final Log logger = LogFactory.getLog(SerializerHelper.class);
  /**
   * The singleton instance of the serialize helper.
   */
  private static SerializerHelper singleton;

  /**
   * Returns or creates a new SerializerHelper. When a new instance is created
   * by this method, all known SerializeMethods are registered.
   *
   * @return the SerializerHelper singleton instance.
   */
  public static synchronized SerializerHelper getInstance()
  {
    if (singleton == null)
    {
      singleton = new SerializerHelper();
      singleton.registerMethods();
    }
    return singleton;
  }


  /**
   * This method can be used to replace the singleton instance of this helper.
   *
   * @param helper the new instance of the serialize helper.
   */
  protected static void setInstance(final SerializerHelper helper)
  {
    singleton = helper;
  }

  /**
   * A collection of the serializer methods.
   */
  private final HashMap methods;

  /**
   * A class comparator for searching the super class of an certain class.
   */
  private final ClassComparator comparator;

  /**
   * Creates a new SerializerHelper.
   */
  protected SerializerHelper()
  {
    this.comparator = new ClassComparator();
    this.methods = new HashMap();
  }

  /**
   * Registers a new SerializeMethod with this SerializerHelper.
   *
   * @param method the method that should be registered.
   */
  public synchronized void registerMethod(final SerializeMethod method)
  {
    this.methods.put(method.getObjectClass(), method);
  }

  /**
   * Traverses the configuration and registers all serialization handlers in this factory.
   */
  protected void registerMethods()
  {
    final Configuration config = LibSerializerBoot.getInstance().getGlobalConfig();
    final Iterator sit = config.findPropertyKeys("org.pentaho.reporting.libraries.serializer.handler.");

    while (sit.hasNext())
    {
      final String configkey = (String) sit.next();
      final String c = config.getConfigProperty(configkey);
      final Object maybeModule = ObjectUtilities.loadAndInstantiate
          (c, SerializerHelper.class, SerializeMethod.class);
      if (maybeModule != null)
      {
        final SerializeMethod module = (SerializeMethod) maybeModule;
        registerMethod(module);
      }
      else
      {
        logger.warn("Invalid SerializeMethod implementation: " + c);
      }
    }
  }

  /**
   * Deregisters a new SerializeMethod with this SerializerHelper.
   *
   * @param method the method that should be deregistered.
   */
  public synchronized void unregisterMethod(final SerializeMethod method)
  {
    this.methods.remove(method.getObjectClass());
  }

  /**
   * Returns the collection of all registered serialize methods.
   *
   * @return a collection of the registered serialize methods.
   */
  protected HashMap getMethods()
  {
    return methods;
  }

  /**
   * Returns the class comparator instance used to find correct super classes.
   *
   * @return the class comparator.
   */
  protected ClassComparator getComparator()
  {
    return comparator;
  }

  /**
   * Looks up the SerializeMethod for the given class or null if there is no
   * SerializeMethod for the given class.
   *
   * @param c the class for which we want to lookup a serialize method.
   * @return the method or null, if there is no registered method for the
   *         class.
   */
  protected SerializeMethod getSerializer(final Class c)
  {
    final SerializeMethod sm = (SerializeMethod) methods.get(c);
    if (sm != null)
    {
      return sm;
    }
    return getSuperClassObjectDescription(c);
  }

  /**
   * Looks up the SerializeMethod for the given class or null if there is no
   * SerializeMethod for the given class. This method searches all
   * superclasses.
   *
   * @param d the class for which we want to lookup a serialize
   *          method.
   * @return the method or null, if there is no registered method for the
   *         class.
   */
  protected SerializeMethod getSuperClassObjectDescription
      (final Class d)
  {
    SerializeMethod knownSuperClass = null;
    final Iterator keys = methods.keySet().iterator();
    while (keys.hasNext())
    {
      final Class keyClass = (Class) keys.next();
      if (keyClass.isAssignableFrom(d))
      {
        final SerializeMethod od = (SerializeMethod) methods.get(keyClass);
        if (knownSuperClass == null)
        {
          knownSuperClass = od;
        }
        else
        {
          if (comparator.isComparable
              (knownSuperClass.getObjectClass(), od.getObjectClass()))
          {
            if (comparator.compare
                (knownSuperClass.getObjectClass(), od.getObjectClass()) < 0)
            {
              knownSuperClass = od;
            }
          }
        }
      }
    }
    return knownSuperClass;
  }


  /**
   * Writes a serializable object description to the given object output stream.
   * This method selects the best serialize helper method for the given object.
   *
   * @param o   the to be serialized object.
   * @param out the outputstream that should receive the object.
   * @throws IOException if an I/O error occured.
   */
  public synchronized void writeObject(final Object o,
                                       final ObjectOutputStream out)
      throws IOException
  {
    try
    {
      if (o == null)
      {
        out.writeByte(0);
        return;
      }
      if (o instanceof Serializable)
      {
        out.writeByte(1);
        out.writeObject(o);
        return;
      }

      final SerializeMethod m = getSerializer(o.getClass());
      if (m == null)
      {
        throw new NotSerializableException(o.getClass().getName());
      }
      out.writeByte(2);
      out.writeObject(m.getObjectClass());
      m.writeObject(o, out);
    }
    catch (NotSerializableException nse)
    {
      DebugLog.log("Unable to serialize object: " + o);
      throw nse;
    }
  }

  public synchronized boolean isSerializable(final Object o)
  {
    if (o == null)
    {
      return true;
    }
    if (o instanceof Serializable)
    {
      return true;
    }

    final SerializeMethod m = getSerializer(o.getClass());
    return m != null;
  }

  /**
   * Reads the object from the object input stream. This object selects the best
   * serializer to read the object.
   * <p/>
   * Make sure, that you use the same configuration (library and class versions,
   * registered methods in the SerializerHelper) for reading as you used for
   * writing.
   *
   * @param in the object input stream from where to read the serialized data.
   * @return the generated object.
   * @throws IOException            if reading the stream failed.
   * @throws ClassNotFoundException if serialized object class cannot be found.
   */
  public synchronized Object readObject(final ObjectInputStream in)
      throws IOException, ClassNotFoundException
  {
    final int type = in.readByte();
    if (type == 0)
    {
      return null;
    }
    if (type == 1)
    {
      return in.readObject();
    }
    final Class c = (Class) in.readObject();
    final SerializeMethod m = getSerializer(c);
    if (m == null)
    {
      throw new NotSerializableException(c.getName());
    }
    return m.readObject(in);
  }
}


source/org/pentaho/reporting/libraries/serializer/LibSerializerInfo.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Object Refinery Limited, Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.serializer;

import org.pentaho.reporting.libraries.base.versioning.ProjectInformation;

/**
 * Creation-Date: 23.10.2005, 18:49:39
 *
 * @author Thomas Morgner
 */
public class LibSerializerInfo extends ProjectInformation
{
  /**
   * The info singleton.
   */
  private static LibSerializerInfo singleton;

  /**
   * Returns the single instance of this class.
   *
   * @return The single instance of information about the JCommon library.
   */
  public static synchronized LibSerializerInfo getInstance()
  {
    if (singleton == null)
    {
      singleton = new LibSerializerInfo();
      singleton.initialize();
    }
    return singleton;
  }


  /**
   * Constructs an empty project info object.
   */
  private LibSerializerInfo()
  {
    super("libserializer", "LibSerializer");
  }

  /**
   * Second step of the initialization.
   */
  private void initialize() 
  {
    setInfo("http://reporting.pentaho.org/libserializer/");
    setCopyright("(C)opyright 2006-2010, by Pentaho Corporation, Object Refinery Limited and Contributors");

    setLicenseName("LGPL");

    setBootClass(LibSerializerBoot.class.getName());
  }

}
